File: algorithms/2_medium/K/01191_k_concatenation_maximum_sum/solution.c
#define MOD 1000000007

static inline int64_t max_i64(const int64_t a, const int64_t b)
{
    return a > b ? a : b;
}

int kConcatenationMaxSum(const int* arr, const int len, const int k)
{
    assert(k >= 1 && len > 0 && arr != NULL);
    int64_t max_sub = 0;
    int64_t curr = 0;
    for (int i = 0; i < len; i += 1) {
        curr += arr[i];
        max_sub = max_i64(max_sub, curr);
        if (curr < 0) {
            curr = 0;
        }
    }

    if (k == 1 || max_sub == 0) {
        return (int)(max_sub % MOD);
    }

    int64_t sum = 0;
    for (int i = 0; i < len; i += 1) {
        sum += arr[i];
    }
    sum = max_i64(sum, 0);

    int64_t max_l = 0;
    curr = 0;
    for (int i = 0; i < len; i += 1) {
        curr += arr[i];
        max_l = max_i64(max_l, curr);
    }

    int64_t max_r = 0;
    curr = 0;
    for (int i = len - 1; i >= 0; i -= 1) {
        curr += arr[i];
        max_r = max_i64(max_r, curr);
    }

    const int64_t res = max_i64(sum * (k - 2) + max_l + max_r, max_sub);
    return (int)(res % MOD);
}

File: algorithms/2_medium/K/01191_k_concatenation_maximum_sum/solution.rs
impl Solution {
    const MOD: i64 = 1_000_000_007;

    pub fn k_concatenation_max_sum(arr: Vec<i32>, k: i32) -> i32 {
        assert!(k >= 1 && arr.len() > 0);
        let k = k as i64;
        let mut max_sub = 0_i64;
        let mut curr = 0_i64;
        for &a in &arr {
            curr += a as i64;
            max_sub = max_sub.max(curr);
            if curr < 0 {
                curr = 0;
            }
        }

        if k == 1 || max_sub == 0 {
            return (max_sub % Self::MOD) as i32;
        }

        let sum = arr.iter().map(|x| *x as i64).sum::<i64>().max(0_i64);

        let mut max_l = 0_i64;
        let mut curr = 0_i64;
        for &a in &arr {
            curr += a as i64;
            max_l = max_l.max(curr);
        }

        let mut max_r = 0_i64;
        let mut curr = 0_i64;
        for &a in arr.iter().rev() {
            curr += a as i64;
            max_r = max_r.max(curr);
        }

        ((sum * (k - 2) + max_l + max_r).max(max_sub) % Self::MOD) as i32
    }
}
